File: leet/leet_560.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int len = nums.size();
        map<int, int> sum_cnt;
        ++sum_cnt[0]; // 1: 0 is a natural sum
        int sum = 0, cnt = 0;
        for(int i = 0; i < len; ++i)
        {
            sum += nums[i];
            //check if sum-k exist: map.count return 1 of exist, 0 otherwise
            if(sum_cnt.count(sum - k) > 0)
            {
                cnt += sum_cnt.at(sum - k);
            }
            ++sum_cnt[sum];
        }
        return cnt;
    }
};

int main()
{
	int a[] = {1,1,1};
	vector<int> arr(a, a + sizeof(a)/sizeof(a[0]));
	Solution solu;
	cout << solu.subarraySum(arr, 2) << endl;
	return 0;
}

File: leet/leet_560.py
# find total number of continous subarrays whose sum is k
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        leng = len(nums)
        sum_cnt = dict()
        # 0 is natural sum factor
        sum_cnt[0] = 1
        sum_tmp = 0
        cnt = 0
        for i in range(leng):
            # integral (sum) from [0] to [i]
            sum_tmp += nums[i]
            if sum_cnt.has_key(sum_tmp - k):
                 cnt += sum_cnt[sum_tmp - k]
            if sum_cnt.has_key(sum_tmp):
                 sum_cnt[sum_tmp] += 1
            else:
               sum_cnt[sum_tmp] = 1
        return cnt

if __name__ == '__main__':
    solu = Solution()
    print solu.subarraySum([1,1,1], 2) 
